processing/src/main/java/org/apache/druid/frame/channel/ReadableInputStreamFrameChannel.java

```diff
@@ -186,7 +190,15 @@ private void startReading()
                 totalInputStreamBytesRead += bytesRead;
                 if (backpressureFuture != null) {
                   keepReading = false;
-                  backpressureFuture.addListener(() -> keepReading = true, Execs.directExecutor());
+                  backpressureFuture.addListener(
+                      () -> {
+                        keepReading = true;
+                        synchronized (readMonitor) {
```

Comment:
There's a potential race here where:

1) The reader thread sees `!keepReading` on line 153
2) `keepReading` is set to `true` here
3) readMonitor is notified here
4) The reader thread enters the `synchronized (readMonitor)` block on line 157, and waits for no reason

We can fix it by setting `keepReading` inside this block, and double-checking it in the reader thread on entry into the `synchronized (readMonitor)` block.

Reply:
Thanks for catching it Gian! Fixed it